store/sqlstore/upgrade.go

```diff
@@ -1208,6 +1209,201 @@ func upgradeDatabaseToVersion537(sqlStore *SqlStore) {
+		updateReplyCountsQuery = `
+			UPDATE Threads
+			INNER JOIN (` + countsAndLastReplyCTE + `) as q
+			ON Threads.PostId = q.RootId
+			SET ReplyCount = q.FixedReplyCount, LastReplyAt = q.FixedLastReply
+		`
+	}
+
+	participantsCTE := `
+		SELECT COALESCE(NULLIF(RootId,''), Id) AS ThreadId, ARRAY_AGG(DISTINCT UserId) FixedParticipants
+		FROM Posts
+		WHERE Posts.DeleteAt = 0
+		GROUP BY ThreadId
+	`
+	updateParticipantsQuery := `
+		WITH q AS (` + participantsCTE + `)
+		UPDATE Threads
+		SET Participants = CONCAT('["', ARRAY_TO_STRING(q.FixedParticipants,'","'), '"]')
+		FROM q WHERE Threads.PostId = q.ThreadId
+	`
+	if sqlStore.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		updateParticipantsQuery = `
+			UPDATE Threads
+			INNER JOIN (SELECT COALESCE(NULLIF(RootId,''), Id) AS ThreadId, GROUP_CONCAT(DISTINCT UserId SEPARATOR '\", \"') AS FixedParticipants FROM Posts WHERE Posts.DeleteAt = 0 GROUP BY ThreadId) as q
+			ON Threads.PostId = q.ThreadId
+			SET Participants = CONCAT('[\"', q.FixedParticipants, '\"]')
+                `
+	}
+
+	now := fmt.Sprintf("%d", model.GetMillis())
+	threadMembershipsCTE := `
+		SELECT PostId, UserId, ChannelMembers.LastViewedAt as CM_LastViewedAt, Threads.LastReplyAt
+		FROM Threads
+		INNER JOIN ChannelMembers on ChannelMembers.ChannelId = Threads.ChannelId
+		WHERE Threads.LastReplyAt <= ChannelMembers.LastViewedAt
+	`
+	updateThreadMembershipQuery := `
+		WITH q as (` + threadMembershipsCTE + `)
+		UPDATE ThreadMemberships set LastViewed = q.CM_LastViewedAt, UnreadMentions = 0, LastUpdated =` + now + `
+		FROM q WHERE ThreadMemberships.Postid = q.PostId AND ThreadMemberships.UserId = q.UserId
+	`
+	if sqlStore.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		updateThreadMembershipQuery = `
+			UPDATE ThreadMemberships
+			INNER JOIN (` + threadMembershipsCTE + `) as q
+			ON ThreadMemberships.Postid = q.PostId AND ThreadMemberships.UserId = q.UserId
+			SET LastViewed = q.CM_LastViewedAt, UnreadMentions = 0, LastUpdated =` + now + `
+		`
+	}
+
+	if _, err := sqlStore.GetMaster().ExecNoTimeout(updateReplyCountsQuery); err != nil {
+		mlog.Error("Error updating reply counts and lastreplyat of threads", mlog.Err(err))
+	}
+	if _, err := sqlStore.GetMaster().ExecNoTimeout(updateParticipantsQuery); err != nil {
+		mlog.Error("Error updating participants of threads", mlog.Err(err))
+	}
+	if _, err := sqlStore.GetMaster().ExecNoTimeout(updateThreadMembershipQuery); err != nil {
+		mlog.Error("Error updating lastviewedat and unreadmentions of threadmemberships", mlog.Err(err))
+	}
+}
+
+// fixCRTChannelMembershipCounts fixes the channel counts, i.e. the total message count,
+// total root message count, mention count, and mention count in root messages for users
+// who have viewed the channel after the last post in the channel
+func fixCRTChannelMembershipCounts(sqlStore *SqlStore) {
+	now := fmt.Sprintf("%d", model.GetMillis())
+
+	channelMembershipsCountsAndMentions := `
```

Comment:
We need to parameterize this and the query below to avoid SQL injections. I know the input is in our control, but it's a good practice and keeps the code clean.